---
title: "BIOST 578 - Final Project - RNA-seq analysis with DER Finder"
author: "Kelsey Grinde, Yatong Li, Xiaowen Tian"
date: "3/5/2015"
output: slidy_presentation
---

## Differential expression analysis of RNA-seq data at single-base resolution

Authors: Alyssa Frazee, Sarven Sabunciyan, Kasper Hansen, Rafael Irizarry and Jeffrey Leek

Outline of talk:

- Introduction/motivations
- Overview of methods: linear models, HMMs, permutation tests
- Reproduction of some of the paper's results
- Discussion/questions

## Introduction

- introduce context and motivations for paper
- relate to what we learned in class

## Methods - Outline

1. Quantify DE at each base using linear model framework
2. Identify DERs using HMM
3. Assess significance of DERs using permutation test

## Methods - Quantify DE at each base using linear model framework

- explain methods (show the equation, can use latex syntax, e.g., $g(x)=\beta$)
- relate to LIMMA

## Methods - Quantify DE at each base using linear model framework
- A coverage matrix has rows corresponding to bases and columns corresponding to smaples.
- Each entry $Y_{ij}$ of the coverage matrix is the number of aligned reads from from sample $i$ that overlap base $j$.
- We fit a linear regression model to each row of the coverage matrix: $$g(Y_{ij}) = \alpha(l_j) + \beta(l_j)X_{pi} + \sum_{k=1}^K \gamma_k(l_j)W_{ik} + \epsilon_{ij}$$

  
## Methods - Quantify DE at each base using linear model framework (cont.)
- $$g(Y_{ij}) = \alpha(l_j) + \beta(l_j)X_{pi} + \sum_{k=1}^K \gamma_k(l_j)W_{ik} + \epsilon_{ij}$$
- $\alpha(l_j)$: baseline gene expression (coverage)
- $X_{i}$: indicator of whether sample $i$ is a case or a control
- $\beta(l_j)$: parameter of interest quantifying differential expression between case and control at location $l_j$.
- $W_{ik}$: values of potential confounders for sample $i$ (e.g. GC content effect).
- $\epsilon_{ij}$: residual measurement error at location $j$.

## Methods - Identify DERs using HMM

- overview of concepts/basic ideas/purpose (maybe avoid technical details? or go into only if time allows?)

## Methods - Assess significance of DERs using permutation test

- explain concepts behind permutation test
- briefly describe methods

## Results

comparing DER Finder to other methods:
- Y chromsome example data

## Discussion

Advantages and limitations of this approach

## Thanks!
Many thanks to the original authors for their documentation, code, and data on github that made it much easier to reproduce their results! 
Thanks also to Raphael and Brian for giving us the knowledge and resources necessary to complete this project.

## Questions?

